### ISSUE TITLE
Version 0.1 todo

### ISSUE THREAD

**Issue by c42f** [JuliaGeo/Proj4.jl#2]
Here's an issue for discussing what needs to happen before registering the package.  Things to do:

  - [x] Settle on a name.  Candidates seem to include: LibProj.jl, Proj4.jl, Proj.jl
  - [x] Consider what's missing or inconsistent as basic functionality in the public API, as compared to other wrappers (pyproj, proj4js, others?).
  - [x] Create build.jl to install libproj on osx, linux (apt, yum?), windows.
  - [ ] Basic documentation

Nice but not strictly necessary:
  - [x] Definitions for travis and appveyor CI

**Comment by yeesian** [comment]
> Settle on a name. Candidates seem to include: LibProj.jl, Proj4.jl, Proj.jl

Your call (:

> Consider what's missing or inconsistent as basic functionality in the public API, as compared to other wrappers (pyproj, proj4js, others?).

I'd say the basic functionality is already there. While I have momentum in this package, I don't mind wrapping the rest of the functions in the C API (partly also because I have use for them in other packages) over this weekend, without having to tag and re-tag in METADATA.

cc @garborg

**Comment by c42f** [comment]
Hah, naming is hard!  Looking around at different packages, there actually seems to be more or less a consensus that a 4 should be included in the name:

https://github.com/OSGeo/proj.4 (the original)
https://github.com/Caged/proj4rb
https://github.com/pebbe/go-proj-4
https://github.com/jswhit/pyproj
https://github.com/proj4js/proj4js (a port, not a wrapper)

**Comment by c42f** [comment]
So let's go with Proj4.jl

**Comment by yeesian** [comment]
:+1:

**Comment by yeesian** [comment]
> Consider what's missing or inconsistent as basic functionality in the public API, as compared to other wrappers (pyproj, proj4js, others?).

Looking through https://github.com/OSGeo/proj.4/blob/master/src/proj_api.h and https://github.com/OSGeo/proj.4/blob/master/src/geodesic.h,

May I propose the following (public) methods?

```julia
pj_transform
pj_geocentric_to_geodetic
pj_geodetic_to_geocentric
pj_compare_datums
pj_is_latlong
pj_is_geocent
pj_get_spheroid_defn
pj_get_def
pj_latlong_from_proj

geod_direct
geod_inverse
```

**Comment by c42f** [comment]
I'm certainly open to exposing the low level C API, but I'd like to have some specific reason for it.  Do you have a use case in mind?

**Comment by yeesian** [comment]
> I'm certainly open to exposing the low level C API, but I'd like to have some specific reason for it. Do you have a use case in mind?

I can't be concrete without already wrapping the C API, but here's one: [OSM data is usually provided in geodetic coordinates](http://gis.stackexchange.com/questions/48949/epsg-3857-or-4326-for-googlemaps-openstreetmap-and-leaflet).

When constructing the road network, it'd be nice to have a default (WGS84) reference ellipsoid to compute the distances on, rather than requiring the user to specify a 2D local plane for the points to be projected onto (the way it is currently done in OpenStreetMap.jl now). LibGEOS only performs operations on the cartesian plane, so the `geod_inverse` method (in Proj4) is required.

The `geod_inverse` method requires the parameters of the ellipsoid (of any given projection), hence the `pj_get_spheroid_defn` is required. `pj_compare_datums` is a neater way of comparing datums than what I currently do in the tests (which is to repeatedly alternate between `pj_get_def` and `pj_transform` until I hit a fixed point). `pj_geocentric_to_geodetic` and `pj_geodetic_to_geocentric` will be required when plotting map data on a 2D graphics library. Everything else listed has already been exposed, with the exception of `pj_latlong_from_proj`, which sounds like a reasonable thing to ask for.

**Comment by c42f** [comment]
Sure, I'd like to have the *functionality* from all that, but it seems like we could do better than exposing the low level C API publicly?

**Comment by yeesian** [comment]
Oh, I see now how misleading my previous messages were. I meant public convenience methods that are based on calls to these C functions, rather than having these low-level wrappers as public methods. Sorry for the noise.

**Comment by c42f** [comment]
No problems at all!

**Comment by c42f** [comment]
I've merged your branch and I really must leave this alone for a few hours now!  If you want to tackle the build stuff (since I can see you've already got some experience there) that would be great, but feel free to have a go at anything else of course.  Build-wise I can only easily test on linux, though I do have a windows 7 VM which could be set up if necessary (osx is unfortunately completely off my radar).

**Comment by yeesian** [comment]
I think I've covered whatever's considered

> basic functionality in the public API

in PR #4. Feel free to comment there, if you find anything missing that you would like to see in the first release.

**Comment by c42f** [comment]
Well, this issue has sat around for a stupidly long time (my fault entirely).

At this stage, I'm afraid the API is going to change drastically in the relatively near future, but several people have already used the package for actual work, which means I need to just register and tag it regardless!  That way at least they can work against a known tagged version.

**Comment by c42f** [comment]
Rather than leaving this hanging any longer, I'm registering Proj4 at the current version.  It's been stable (if not finished) for a long time now.

Ref. https://github.com/JuliaLang/METADATA.jl/pull/6450
